test/e2e/cluster_upgrade.go

```diff
@@ -36,68 +35,176 @@ import (
 	. "github.com/onsi/gomega"
 )
 
+// version applies to upgrades; kube-push always pushes local binaries.
+const version = "latest_ci"
+
+// The following upgrade functions are passed into the framework below and used
+// to do the actual upgrades.
+
+var masterUpgrade = func() error {
+	_, _, err := runScript("hack/e2e-internal/e2e-upgrade.sh", "-M", version)
+	return err
+}
+
+var masterPush = func() error {
+	_, _, err := runScript("hack/e2e-internal/e2e-push.sh", "-m")
+	return err
+}
+
+var nodeUpgrade = func(f Framework, replicas int) error {
+	Logf("Preparing node upgarde by creating new instance template")
+	stdout, _, err := runScript("hack/e2e-internal/e2e-upgrade.sh", "-P", version)
+	if err != nil {
+		return err
+	}
+	tmpl := strings.TrimSpace(stdout)
+
+	Logf("Performing a node upgrade to %s; waiting at most %v per node", tmpl, restartPerNodeTimeout)
+	if err := migRollingUpdate(tmpl, restartPerNodeTimeout); err != nil {
```

Comment:
But if the bash changes (or breaks) we won't notice because it isn't covered by the test. The point of the test isn't to replicate the upgrade functionality more reliably, but rather to exercise the functionality. 

I'm ok with this going in as-is if we a) add a big fat comment to the one line in the bash script saying not to change it without updating this test and b) we file an issue to replace the bash script with a go program that we can reasonably test.